R/colors.R

```diff
@@ -0,0 +1,315 @@
+#'   then the result will contain `NA` for invalid colors.
+#'
+#' @examples
+#'
+#' parseCssColors(c(
+#'   "#0d6efd",
+#'   "#DC35457F",
+#'   "rgb(32,201,151)",
+#'   "  rgba( 23 , 162 , 184 , 0.5 )  ",
+#'   "hsl(261, 51%, 51%)",
+#'   "cornflowerblue"
+#' ))
+#' @export
+parseCssColors <- function(str, mustWork = TRUE) {
+  # Strip insignificant whitespace
+  str <- color_strip_ws(str)
+
+  strategies <- list(
+    # #RRGGBBAA and #RRGGBB
+    list(
+      pattern = "^#([[:xdigit:]]{2})([[:xdigit:]]{2})([[:xdigit:]]{2})([[:xdigit:]]{2})?$",
+      decoders = list(
+        decode_hex, # red
+        decode_hex, # green
+        decode_hex, # blue
+        decode_optional(decode_hex, 0xFF) # alpha, optional
+      ),
+      encoder = encode_hex
+    ),
+    # #RGBA and #RGB
+    list(
+      pattern = "^#([[:xdigit:]])([[:xdigit:]])([[:xdigit:]])([[:xdigit:]])?$",
+      decoders = list(
+        decode_hex, # red
+        decode_hex, # green
+        decode_hex, # blue
+        decode_optional(decode_hex, 0xFF) # alpha, optional
+      ),
+      encoder = encode_hex
+    ),
+    # rgb() and rgba()
+    list(
+      pattern = paste0("^rgba?\\(",
+        "(", regex_float, "),",
+        "(", regex_float, "),",
+        "(", regex_float, ")",
+        regex_non_capturing_group(",(", regex_float, ")"), "?",
+        "\\)$"),
+      decoders = list(
+        decode_float_255, # red   [0-255]
+        decode_float_255, # green [0-255]
+        decode_float_255, # blue  [0-255]
+        decode_optional(decode_float_1, 0xFF) # alpha [0-1], optional
+      ),
+      encoder = encode_hex
+    ),
+    # hsl() and hsla()
+    list(
+      pattern = paste0("^hsla?\\(",
+        "(", regex_float, "),",
+        "(", regex_float, ")%,",
+        "(", regex_float, ")%",
+        regex_non_capturing_group(",(", regex_float, ")"), "?",
+        "\\)$"),
+      decoders = list(
+        decode_float_identity, # hue        [0-360]
+        decode_float_identity, # saturation [0-255]
+        decode_float_identity, # lightness  [0-255]
+        decode_optional(decode_float_1, 0xFF) # alpha [0-1], optional
+      ),
+      encoder = encode_hsl
+    ),
+    # color keywords
+    list(
+      pattern = paste0("^", regex_color_keyword, "$"),
+      decoders = list(decode_color_keyword),
+      encoder = encode_hex
+    )
+  )
+
+  success <- rep_len(FALSE, length(str))
+  result <- rep_len(NA_character_, length(str))
+
+  for (strat in strategies) {
+    if (all(success)) {
+      break
+    }
+
+    res <- match_and_decode(
+      str[!success],
+      strat$pattern,
+      !!!strat$decoders
+    )
+    if (any(res$matching_rows)) {
+      result[!success][res$matching_rows] <- strat$encoder(res$values)
+      success[!success][res$matching_rows] <- TRUE
+    }
+  }
+
+  if (mustWork && any(!success)) {
+    stop("CSS color value(s) could not be parsed")
+  }
+
+  result
+}
+
+# Strips whitespace that isn't significant in the parsing of CSS colors.
```

Comment:
What's an example of significant white-space (I don't immediately see why a naive approach to stripping white-space doesn't work)

Reply:
Here are some examples that *shouldn't* parse:

```
"# FF FF FF"
"rgb (64, 64, 64)"      # space between "rgb" and "(" isn't allowed
"hsl(180, 50 %, 50 %)"  # space before % isn't allowed
```